### DISCUSSION TITLE
How to fix DDLogXYZ deprecation (using interpolated DDLogMessageFormat instead)?

### DISCUSSION THREAD

**Discussion by AntonDe** [CocoaLumberjack/CocoaLumberjack#1394] [Q&A] [answered]
Hi there,

I've some warnings in my project where I use the DDLogInfo macro, with an suggestion to use DDLogMessageFormat:

`
'DDLogInfo(_:level:context:file:function:line:tag:asynchronous:ddlog:)' is deprecated: Use an interpolated DDLogMessageFormat instead
`

This warning occurs only in Swift code (I have both: Objective C and Swift code) and only at parameterized messages.
The suggested DDLogMessageFormat struct takes an stringLiteral or stringInterpolation as a parameter, but I'm not able to set the log level. 
I'm found a solution to create an DDLogMessage instance and pass it to DDLog.log but is there a simpler way to achieve this and fix the warnings.

Best thanks

**Comment by ffried** [comment, marked as answer]
Simply use a Swift string literal for logging messages:

```swift
DDLogInfo("My log message")
```

If you have parameters in your log message, use a Swift string interpolation:

```swift
let parameter = "My Parameter"
DDLogInfo("My log message with parameter: \(parameter)")
```

<details>
<summary><b>Old vs. new examples</b></summary>

If you've previously had a string as the message in a separate variable, inline it into the `DDLogXYZ` call. Alternatively, declare it as a `DDLogMessageFormat`. Note however, that the latter will always evaluate the message even if won't be logged due to the configured level.

**Old:**
```swift
let message = "My message"
DDLogInfo(message)
```

**New:**
```swift
DDLogInfo("My message")
// OR (always evaluating the message)
let message: DDLogMessageFormat = "My message"
DDLogInfo(message)
```

<hr/>

If you have previously passed an arbitrary value to the `DDLogXYZ` functions, you need to wrap them in a string interpolation now. Note, though, that it's recommended to form a real message at the call site of `DDLogXYZ`.

**Old:**
```swift
let myValue = getSomeValue()
DDLogInfo(myValue)
```

**New:**
```swift
let myValue = getSomeValue()
DDLogInfo("\(myValue)")
```
</details>


<details>
<summary><b>Wrapper examples</b></summary>
If you're using a wrapper around the `DDLogXYZ` functions, make sure their message parameter is now of type `DDLogMessageFormat`. After that, the same rules apply for your wrapper as with `DDLogXYX` explained above.

**Old:**
```swift
func myInfoLogWrapper(_ message: Any) {
    DDLogInfo(message)
}
```

**New:**
```swift
func myInfoLogWrapper(_ message: DDLogMessageFormat) {
    DDLogInfo(message)
}
```
<details>

**Comment by Sega-Zero** [comment]
having `'DDLogInfo(_:level:context:file:function:line:tag:asynchronous:ddlog:)' is deprecated: Use an interpolated DDLogMessageFormat instead` when using this way:
```swift
DDLogInfo("some logging string \(somevar), \(Bundle.main.bundleIdentifier, default: "com.company.id") \(someothervar)")
```
the warning is not produced if using it this way:
```swift
DDLogInfo("some logging string \(somevar), \(Bundle.main.bundleIdentifier ?? "com.company.id") \(someothervar)")
```

@ffried should I make a separate issue?

**Reply by ffried** [reply]
The `\(optionalXYZ, default: "some string")` interpolation is a new feature in Swift 6.2 for the default string interpolation. Thus when using it, it results in a `String` rather than a `DDLogMessageFormat`.
I'll see if I can add support for this in `DDLogMessageFormat` as well.

**Reply by ffried** [reply]
This has landed on `master` and will be in the next release.
